runtime/executor/src/main/java/org/apache/nemo/runtime/executor/transfer/LocalInputContext.java

```diff
@@ -0,0 +1,113 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.nemo.runtime.executor.transfer;
+
+import org.apache.nemo.common.punctuation.Finishmark;
+import java.util.Iterator;
+import java.util.concurrent.LinkedBlockingQueue;
+
+/**
+ * This class provides a data transfer interface to the receiver side when both the sender and the receiver are in the
+ * same executor. Since the sender doesn't serialize data, the receiver doesn't need to deserialize data when retrieving
+ * them.
+ */
+  public final class LocalInputContext extends LocalTransferContext {
+    private final LinkedBlockingQueue queue;
+    private boolean isClosed = false;
+
+    /**
+     * Creates a new local input context and connect it to {@param localOutputContext}.
+     * @param localOutputContext the local output context to which this local input context is connected
+     */
+    public LocalInputContext(final LocalOutputContext localOutputContext) {
+    super(localOutputContext.getExecutorId(),
+          localOutputContext.getEdgeId(),
+          localOutputContext.getSrcTaskIndex(),
+          localOutputContext.getDstTaskIndex());
+    this.queue = localOutputContext.getQueue();
+  }
+
+  /**
+   * Closes this local input context.
+   * @throws RuntimeException if the connected output context hasn't been closed yet, or if there are still data
+   * left to be processed.
+   */
+  @Override
+  public void close() throws RuntimeException {
+    isClosed = true;
+  }
+
+  /**
+   * Checks if this context has already been closed.
+   * @return true if this context has already been closed.
+   */
+  public boolean isClosed() {
+    return isClosed;
+  }
+
+  /**
+   * Creates a new iterator which iterates the receive elements from the sender.
+   * @return iterator that iterates the received elements.
+   */
+  public LocalInputIterator getIterator() {
+    return new LocalInputIterator();
+  }
+
+  /**
+   * Local input iterator that iterates the received elements from the sender.
+   */
+  private class LocalInputIterator implements Iterator {
+    private Object next;
+    private boolean hasNext = false;
+
+    @Override
+    public final boolean hasNext() {
+      if (hasNext) {
+        return true;
+      }
+      if (isClosed) {
+        return false;
+      }
+      try {
+        // Blocking call
+        next = queue.take();
+        if (next instanceof Finishmark) {
```

Comment:
Do we need `isClosed` variable? if we check finishmark, we don't need `isClosed` variable.

Reply:
isClosed variable handles the case when the "hasNext" method gets called multiple times while the "next" method doesn't get invoked. I'm aware that this case doesn't happen since the data fetcher calls the "next" method every time the "hasNext" method returns true. But to make it more general, we should handle this common case also.